rfcs/20200504-tff-aggregators-placement.md

```diff
@@ -0,0 +1,226 @@
+party to perform the secure shuffling needed to realize a differential privacy
+guarantee. Since this is an established area of the literature with strong
+motivations and results, we see this as an important line of work for TFF to
+support in order to keep with its
+[project goals](https://github.com/tensorflow/federated#tensorflow-federated).
+
+In general, any secure aggregation protocol can be represented as a coordinated
+computation between three groups of parties: a server, a (potentially singleton)
+set of aggregators, and a set of clients. Note that these need not be mutually
+exclusive, so for example the traditional parameter server setting can be recovered
+as a special case by treating the server as a singleton aggregators set.
+
+The TFF Federated Core (FC) language currently realizes logically-distinct parties
+as "placements". While there exist `tff.SERVER` and `tff.CLIENTS` placements in FC,
+there is no `tff.AGGREGATORS` placement. Without such a placement, implementing new
+aggregation protocols in TFF can require low-level programming of the TFF executor
+stacks, as evidenced by
+[this community attempt to integrate secure aggregation](https://github.com/tf-encrypted/rfcs/blob/master/20190924-tensorflow-federated/integration-strategies.md).
+By adding a new `tff.AGGREGATORS` placement, users can more easily implement new
+aggregation protocols by expressing them as federated computations in FC.
+
+## User Benefit
+
+Users can now express custom aggregation protocols in the Federated Core by working
+with federated data placed on `tff.AGGREGATORS`. Users will be unencumbered by the
+constraints of the current federated types in FC.
+
+## Design Proposal
+
+Adding the `tff.AGGREGATORS` placement for federated types involves adding a new
+`Placement` and `PlacementLiteral`, and then extending the compiler to recognize
+federated values with this placement when computing intrinsics. The compiler
+generally defines separate intrinsics by placement; e.g. 
+`tff.federated_value(value, placement)` is actually interpreted by the compiler as
+`federated_value_at_clients(value)` or `federated_value_at_server(value)`,
+depending on the provided `placement`. This means we we will want to add new
+intrinsics that correspond to `tff.AGGREGATORS`, e.g.
+`federated_value_at_aggregators`.
+
+Existing federated computation that will need modification fall into the two
+categories below:
+
+1. Intrinsics for federated computations that are already parameterized by
+placement. Note some of these functions don't have a `placement` arg in their
+public API signature, but internally correspond to different IntrinsicDefs based on
+placement of their federated input(s).
+    - `federated_eval`
+    - `federated_map`
+    - `federated_value`
+    - `federated_zip`
+    - `sequence_map`
+2. Intrinsics that will need to be parameterized by placement, but currently
+aren't. 
+    - `federated_aggregate`
```

Comment:
This will be probably the interesting one. I think the `accumulate` / `merge` / `report` operators should be allowed to be  parameterized by the `tff.AGGREGATORS`-placed values.